src/zelos/zml.py

```diff
@@ -0,0 +1,249 @@
+
+from lark import Lark, Transformer
+
+from zelos.hooks import HookType
+
+
+"""
+ZML or Zelos Manipulation Language is used to specify actions to happen
+on more complex conditions.
+"""
+
+
+class ConditionList:
+    """
+    Specifies common conditions that are used for other ConditionLists
+    """
+
+    def __init__(self, conditions: Dict[str, str]):
+        self._conditions = conditions
+        self._count = 0
+
+    def __str__(self):
+        s = str(self._conditions)
+        return s
+
+    def is_satisfied(self, zelos):
+        thread = self._conditions.get("thread", None)
+        if thread is not None and thread != zelos.thread.name:
+            return False
+
+        # Count must be the last condition
+        self._count += 1
+        n = self._conditions.get("n", None)
+        if n is not None and n != self._count:
+            return False
+        else:
+            return True
+
+
+class SyscallConditionList(ConditionList):
+    """
+    Used to trigger an action after a specific syscall.
+    """
+
+    def __init__(self, conditions: Dict[str, str]):
+        super().__init__(conditions)
+
+    def is_satisfied(self, zelos, sysname, args, retval):
+        if self._conditions["syscall"] != sysname:
+            return False
+
+        retval_condition = self._conditions.get("retval", None)
+        if retval_condition is not None and retval_condition != retval:
+            return False
+
+        arg_conditions = self._conditions.get("arg_", {})
+        for (arg_name, expected_arg_val) in arg_conditions.items():
+            arg_val = getattr(args, arg_name, None)
+            if arg_val is None:
+                # This argument is not present for this syscall.
+                return False
+            if expected_arg_val != arg_val:
+                return False
+
+        return super().is_satisfied(zelos)
+
+    def act_when_satisfied(self, zelos, action: Callable[[], Any]):
+        def trigger_when(zelos, sysname, args, retval):
+            if self.is_satisfied(zelos, sysname, args, retval):
+                action()
+
+        zelos.hook_syscalls(
+            HookType.SYSCALL.AFTER, trigger_when, "on_syscall_conditionlist"
+        )
+
+
+class ApiConditionList(ConditionList):
+    """
+    Used to trigger an action after a specific API.
+    """
+
+    def __init__(self, conditions: Dict[str, str]):
+        super().__init__(conditions)
+
+    def act_when_satisfied(self, zelos, action: Callable[[], Any]):
+        raise NotImplementedError
+
+
+class AddressConditionList(ConditionList):
+    """
+    Used to trigger an action executing a specific address.
+    """
+
+    def __init__(self, conditions: Dict[str, str]):
+        super().__init__(conditions)
+
+    def act_when_satisfied(self, zelos, action: Callable[[], Any]):
+        def trigger_when(zelos, address, size):
+            if self.is_satisfied(zelos):
+                action()
+
+        address = self._conditions["addr"]
```

Comment:
Will this raise a `KeyError` if `self._conditions` does not contain the key name `"addr"`?